### PR TITLE
Modern (as of 2021) rubocop

### PR DESCRIPTION
**Pull request by iHiD**
Our existing rubocop is quite out of date, and the config is an entire rubocop config rather than the list of overrides. This makes it difficult to know what's intentional and what's not, and makes it difficult to bring keep our Ruby up to date with the latest changes/standards. I also think we should have a consistent set of things for our students to read/see/work on.

I'm going through violation-by-violation adding a commit that addresses each. The base file I've started with is what is used elsewhere within Exercism on Ruby projects and matches my preferences. There might be some where people disagree with the Rubocop standard or my preferences. Where that's the case, let's discuss, make a decision, and encode it into the .rubocop file

There will be loads of commits - I'll keep adding them when I have time and I want an "easy" job to be working on.

I also don't know how the generator works for the Ruby track - is a lot of what's getting changed here automatically generated? If so, I guess we need to update the generator to deal with this?

Feel free to comment in the interim, but I'll mark this as not-draft once it's finally ready.

### CODE REVIEW

File: exercises/practice/custom-set/custom_set_test.rb
Comment (kotp) [inline exercises/practice/custom-set/custom_set_test.rb:272]: The position is not important, so this could be reported, but preference is to make the position less important via communication of style. `set1, set2` does this for us, while `set_1` indicates that there is an importance of placement.

Since order does not matter  for union, we should play down the position, using it only for "uniqueness of name".

File: exercises/practice/list-ops/list_ops_test.rb
Comment (kotp) [inline exercises/practice/list-ops/list_ops_test.rb:49]: See importance of position vs importance  of uniqueness communication in prior comment.

Since these are (now or later) generated, the maintainer should make this distinction, if this style is enforced, by indicating in the source that this is purposefully done, via a directive comment.  See https://stackoverflow.com/a/26350879/485864 where I talk about inline directive.

**Review by kotp — COMMENTED** [review]
Some of the comments are likely changed via the `.rubocop.yml` file.

File: .rubocop.yml
Comment (kotp) [inline .rubocop.yml:49]: If we exclude metrics for tests, and pay attention to metrics in others, then we probably don't need to `--except Metrics` when running Rubocop in CI.

File: .rubocop.yml
Comment (kotp) [inline .rubocop.yml:85]: If this is not universally (in the Ruby track) the case, then this should be enabled, and a directive comment can communicate that it is intentional.

Comment (iHiD) [reply .rubocop.yml:85]: Happy to enable it, but then there's a few places that would need fixing and I don't know how to fix them.

Comment (kotp) [reply .rubocop.yml:85]: I can enable it, see what breaks.

Comment (kotp) [reply .rubocop.yml:85]: Feel free to continue to push, I can work through rolling updates, not a problem.

Comment (iHiD) [reply .rubocop.yml:85]: I'm done for now. I need your help on the remaining failures I think.

File: exercises/practice/binary/binary_test.rb
Comment (ErikSchierboom) [inline exercises/practice/binary/binary_test.rb:6]: Is it possible to have Rubocop disable this rule for this file only?

File: exercises/practice/connect/connect_test.rb
Comment (ErikSchierboom) [inline exercises/practice/connect/connect_test.rb:14]: An off-topic remark: I found the way this is written using the map/gsub combination really hard to parse as someone newer to Ruby

**Review by ErikSchierboom — APPROVED** [review]
Big jump in Rubocop versions! Excellent.

**Comment by kotp** [comment]
On 21/02/08 11:45PM, Erik Schierboom wrote:
> @ErikSchierboom approved this pull request.
> > @@ -3,42 +3,42 @@
>  
>  # Common test data version: 1.1.0 dd3c07e
>  class BinaryTest < Minitest::Test
> -  def test_binary_0_is_decimal_0
> +  def test_binary_0_is_decimal_0 # rubocop:disable Naming/VariableNumber
> 
> Is it possible to have Rubocop disable this rule for this file only?

Yes, that is what the comment does.  We would put it at the top of the
file, maybe, and then re-enable at the end.  We can also do so in the
configuration file.

> > @@ -11,7 +11,7 @@ def test_an_empty_board_has_no_winner
>        '  . . . . .',
>        '   . . . . .',
>        '    . . . . .'
> -    ].map {|row| row.gsub(/^ */, '')}
> +    ].map { |row| row.gsub(/^ */, '') }
> 
> An off-topic remark: I found the way this is written using the
> map/gsub combination really hard to parse as someone newer to Ruby

Is it because "new to Ruby" or perhaps "New to regular expressions?"  I
don't disagree though, as I suggest using transliteration (`tr`) instead, or
`scan` instead, usually because scan works by focusing on what is
wanted, rather than getting rid of what you don't want.  A better
viewpoint usually.

**Comment by ErikSchierboom** [comment]
> Is it because "new to Ruby

New to Ruby :)

File: exercises/practice/binary/binary_test.rb
Comment (kotp) [reply exercises/practice/binary/binary_test.rb:6]: It might be better to reword the titles of test methods as well.

File: lib/generator/implementation.rb
Comment (kotp) [inline lib/generator/implementation.rb:38]: This may not even be necessary.  But was not incorrect to use the "no pass through" directive here that is indicated with the `super()`, a quick check indicates the arguments are used locally, and not intended for consumption further up the ancestor chain.  The same quick looking suggests that we would not even need to call the parents `initialize` method.

Comment (kotp) [reply lib/generator/implementation.rb]: Disregard the above for now.  Found the reason for it.

File: exercises/practice/connect/connect_test.rb
Comment (kotp) [reply exercises/practice/connect/connect_test.rb:14]: @ErikSchierboom do you have a way that you would write this instead?  (Getting ready to merge this, pending the failed CI corrections).

**Review by kotp — DISMISSED** [review]

**Comment by iHiD** [comment]
GREEN. Wahooo. Want to Squash + Merge this @kotp?

**Comment by kotp** [comment]
> GREEN. Wahooo. Want to Squash + Merge this @kotp?

Yes, I applied the patch Reading for 3.0, and still green.

I still don't understand why it wants me to do another pull request for Ruby 2.6 and another pull request for Ruby 2.7.

**Review by kotp — APPROVED** [review]

**Comment by kotp** [comment]
> > GREEN. Wahooo. Want to Squash + Merge this @kotp?
To clarify, I can't merge with the:

- Ruby 2.6 - ubuntu-latest - pull_request Expected — Waiting for status to be reported
- Ruby 2.7 - ubuntu-latest - pull_request Expected — Waiting for status to be reported

Status listing as it is.  And I don't know what it means, what it is waiting on…

**Comment by iHiD** [comment]
@kotp All good to go now :)

**Pull request merged**
